### PR TITLE
Low-hanging performance improvements

### PR DESCRIPTION
**Pull request by mdboom**
Instantiating Paths is expensive, but they aren't mutable, so just reuse
when possible.

WeakValueDictionaries are expensive to construct.  We can fix that by
handling the weakrefs ourself.

And `cbook.popall` has no reason to exist.  Maybe someone's using it and it should stay, but no one *should* be using it.

On my benchmark of:

```
plt.subplots(8, 8)
plt.savefig("test.png")
```

I get a speedup from 5.04s to 3.81s.

### CODE REVIEW

**Comment by mdboom** [comment]
I think this is finally ready-to-merge.

File: lib/matplotlib/transforms.py [hunk 1]
```diff
@@ -109,14 +106,17 @@ def __str__(self):
 
     def __getstate__(self):
         d = self.__dict__.copy()
-        # turn the weakkey dictionary into a normal dictionary
-        d['_parents'] = dict(six.iteritems(self._parents))
+        # turn the dictionary with weak values into a normal dictionary
+        d['_parents'] = dict((k, v()) for (k, v) in
```
Comment (tacaswell) [inline lib/matplotlib/transforms.py]: Does this need a check that `v()` returns an active ref?

Comment (mdboom) [inline lib/matplotlib/transforms.py]: I currently check on deserializing (`__setstate__`) where you *have* to deal with it.  No real need to do it in both places unless we *really* care about the size of the serialization.

Comment (tacaswell) [inline lib/matplotlib/transforms.py]: fair enough.

File: lib/matplotlib/markers.py [hunk 2]
```diff
@@ -93,6 +93,8 @@
 (TICKLEFT, TICKRIGHT, TICKUP, TICKDOWN,
  CARETLEFT, CARETRIGHT, CARETUP, CARETDOWN) = list(xrange(8))
 
+_empty_path = Path(np.empty((0, 2)))
```
Comment (pelson) [inline lib/matplotlib/markers.py]: FWIW, there is a fast instantiation route for Paths: https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/path.py#L173.

**Pull request merged**